### ISSUE TITLE
Maintain a list of Column Labels for use in Reporting Pipeline

### ISSUE THREAD

**Issue by jonthegeek** [Gilead-BioStats/gsm.mapping#7]
It would be helpful if this package also included canonical "user-facing" names for every table and field. For example, `subjid` is "SubjectID" in our reports tables, implying that `SubjectID` should be the canonical user-facing name for this field. I'll implement this in gsm.app (probably before this makes it over here), so we can almost definitely migrate code from there.

**Comment by samussiah** [comment]
It shouldn't hurt anything to add/support an optional `label` attribute in data specifications, but I'm not sure where the comprehensive mapped data spec would live... A few options:

- `{gsm.mapping}` generates/maintains the mapped data spec for use by `{gsm.app}`, `{gsm.kri}`, etc.
- `{gsm.template}` generates a mapped data spec alongside a source/raw data spec each time a mapping is updated, checks for new/missing columns, and preserves existing column metadata
- `{gsm.app}` maintains it's own mapped data spec that aligns with the data returned by the mappings in `{gsm.mapping}`
- each mapping can have an optional output data spec

@jwildfire, @lauramaxwell, @jonthegeek, @zdz2101 - thoughts?

**Comment by jwildfire** [comment]
I generally agree that labels are worth adding and I think they almost certainly belong in the `gsm.mapping` YAMLs, but it's not quite trivial since the current `spec` section of the YAMLs really serve as input specifications, not as output specifications.  As such, we can't just add column `labels` as `spec.{domain}.{column}.label` since that won't cover any new columns derived under `steps`. 

I guess we could just add a new `labels` parameter at the top level of the yaml and put something like Jon's objects there ... 

Data set labels can just be added as `meta.label`.

**Comment by jwildfire** [comment]
Another question is how these labels are provided to `gsm.app` and other downstream reports/apps. Unless I'm forgetting something (definitely possible), the workflows aren't really captured in the data model, so there isn't an obvious way to pass these around these without extending the data model - most likely via the reporting framework. I guess that isn't a huge deal, but it would require a few other minor releases (`gsm.reporting` for sure. Maybe in `gsm.template` as well) so it is something to work out before we start implementation here.

**Comment by jwildfire** [comment]
Adding labels to existing YAMLs also needs to deal with how to handle label mismatches. If `ae.subjid.label` doesn't match `pd.subjid.label` then what do we do? Kind of talking myself out of putting labels directly in the YAMLs. 

Maybe we just make an surface object with labels that we share across all domains and allow access where `gsm.mapping::labels$domains$ae` is "Adverse Events" and `gsm.mapping::labels$columns$studyid` is "Study ID". Could add some CI to remind us to update the object when new domains/columns are added. 

To surface labels in modules, we could have `gsm.template` to make a copy of the `labels` object as part of each study build (allowing study level updates as needed), and make the object available in `RunWorkflow` either via `data$labels` or as custom stand-alone `labels` object.

**Comment by jonthegeek** [comment]
I like the separate object idea. That would help us normalize the names across different objects. It's only necessary/useful if a report cares about it, too, so building it into every workflow feels like overkill.

I don't think I grokked how different this could potentially be across studies/companies when I originally wrote this ticket, so I definitely think it'd be a good idea to make this configurable. We'd have a default label object (sent in as `lLabels` probably) for reports that need them, but maybe that default could have a simple way to look for the customization. I was tempted to put that in things like `MakeWorkflowList()` and `Ingest()`, but we don't really need it until/unless we hit a report that cares about it (so only those report functions and probably `RunWorkflows()` would need to know about it).

**Comment by jwildfire** [comment]
I still think we should do this eventually, and that the approach we landed on above sounds feasible. This still sounds about right. 

> Maybe we just make an object with labels that we share across all domains and allow access where gsm.mapping::labels$domains$ae is "Adverse Events" and gsm.mapping::labels$columns$studyid is "Study ID". Could add some CI to remind us to update the object when new domains/columns are added.

**Comment by jwildfire** [comment]
Going to mark this issue a Technical Task since it's really just the creation of a pipeline that will create a `labels` object. When we implement, we'd want to create `feature` issues for each report that would use these standard labels.

**Comment by jonthegeek** [comment]
I still hard-code these translations in gsm.app, and it's annoying when I'm missing them. You can see them more clearly in the Prevalence Plot drop-down menus and axis labels. I'd definitely love to have a more formal way to look them up! I'd add a function that wraps the object while we're at it to make lookups cleaner.

**Comment by jwildfire** [comment]
Yeah, I agree it would be good to standardize before we have to re-invent the wheel in a bunch of reports/apps. I'll pencil it in for Q4, but anticipating it might get bumped to next year.
